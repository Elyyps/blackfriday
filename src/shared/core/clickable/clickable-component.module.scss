$buttonVariants: "primary-default" $global-inverse-color $global-inverse-color,
  "primary-brand" $global-inverse-color $global-inverse-color, "primary-inverted" $gray-dark $gray-dark,
  "secondary" $gray-dark $gray-dark, "secondary-inverted" $gray-dark $gray-dark,
  "tertiary" $gray-dark (darken($brand-primary, 7%) !important),
  "tertiary-inverted" $global-inverse-color $global-inverse-color, "link-primary" $gray-dark $brand-primary,
  "link-secondary" $brand-primary (darken($brand-primary, 7%) !important),
  "link-tertiary" $brand-secondary (darken($brand-secondary, 7%) !important);

@each $name, $color, $colorHover in $buttonVariants {
  .button--#{$name}--icon-filled {
    .icon-svg path {
      fill: $color;
    }
    &:hover .icon-svg path {
      fill: $colorHover;
    }
  }

  .button--#{$name}--icon-outline {
    .icon-svg path {
      stroke: $color;
    }
    &:hover .icon-svg path {
      stroke: $colorHover;
    }
  }

  .button--#{$name}--icon-outline-fill {
    .icon-svg path {
      fill: $color;
      stroke: $color;
    }
    &:hover .icon-svg path {
      fill: $colorHover;
      stroke: $colorHover;
    }
  }
}

a.button,
a.button--primary-default,
.button,
.button--primary-default {
  margin: 0;
  border: none;
  box-sizing: border-box !important;
  -moz-box-sizing: border-box;
  -webkit-box-sizing: border-box;
  font-size: 15px;
  line-height: 1.3;
  font-weight: 600;
  font-family: $global-font-family;
  text-align: center;
  text-decoration: none;
  display: inline-flex;
  align-items: center;
  user-select: none;
  justify-content: center;
  position: relative;
  border-radius: 2px;
  padding: 8px 24px;
  transition: background-color 0.05s;
  cursor: pointer;
  color: $global-inverse-color;
  background-color: $brand-secondary;

  width: 100%;

  &--animated {
    svg {
      transition: transform 150ms ease-in-out;
    }
    &:hover {
      svg {
        transform: translateX(4px);
      }
    }
  }

  .icon-svg {
    display: flex;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  &:hover,
  .button--primary-default:hover {
    text-decoration: none;
    background-color: darken($global-secondary-background, 7%);
  }

  &--primary-brand {
    background-color: $gray-dark;
    &:hover {
      background-color: lighten($gray-dark, 10%);
    }
  }

  &--primary-inverted,
  &--secondary-inverted {
    color: $gray-dark !important;
    background-color: $global-background;
    &:hover {
      background-color: $global-muted-background;
    }
  }

  &--secondary {
    color: $gray-dark !important;
    background-color: $tint-gray-light;
    &:hover {
      background-color: darken($tint-gray-light, 7%);
    }
  }

  &--tertiary {
    color: $gray-dark !important;
    background-color: transparent;
    -webkit-box-shadow: inset 0px 0px 0px 2px $brand-primary;
    -moz-box-shadow: inset 0px 0px 0px 2px $brand-primary;
    box-shadow: inset 0px 0px 0px 2px $brand-primary;
    &:hover {
      -webkit-box-shadow: inset 0px 0px 0px 2px darken($brand-primary, 7%);
      -moz-box-shadow: inset 0px 0px 0px 2px darken($brand-primary, 7%);
      box-shadow: inset 0px 0px 0px 2px darken($brand-primary, 7%);
      color: darken($brand-primary, 7%) !important;
      background-color: lighten($brand-primary, 51%) !important;
    }
  }

  &--tertiary-inverted {
    color: $global-inverse-color;
    background-color: transparent;
    -webkit-box-shadow: inset 0px 0px 0px 2px $brand-primary;
    -moz-box-shadow: inset 0px 0px 0px 2px $brand-primary;
    box-shadow: inset 0px 0px 0px 2px $brand-primary;

    &:hover {
      -webkit-box-shadow: inset 0px 0px 0px 2px darken($brand-primary, 7%);
      -moz-box-shadow: inset 0px 0px 0px 2px darken($brand-primary, 7%);
      box-shadow: inset 0px 0px 0px 2px darken($brand-primary, 7%);
      background-color: darken($brand-primary, 35%) !important;
    }
  }

  &--FAB {
    padding: 0;
    width: 40px;
    height: 40px;
  }

  &--small {
    padding: 8px 18px;
  }

  &--link {
    &-primary,
    &-secondary,
    &-tertiary {
      &,
      &:hover {
        padding: 0;
        background: transparent;
      }
    }

    &-primary {
      color: $gray-dark;

      &:hover {
        color: $brand-primary;
      }
    }
    &-secondary {
      color: $brand-primary;

      &:hover {
        color: darken($brand-primary, 7%) !important;
      }
    }
    &-tertiary {
      color: $brand-secondary;

      &:hover {
        color: darken($brand-secondary, 7%) !important;
      }
    }
  }
}

.icon-left {
  margin-right: 8px;
  display: inherit;
}

.icon-right {
  margin-left: 8px;
  display: inherit;
}

.button__dynamic-size {
  padding: 8px 16px !important;
  background-color: red !important;
}
