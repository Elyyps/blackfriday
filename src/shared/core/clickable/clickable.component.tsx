import React, { useEffect } from "react";
import styles from "./clickable-component.module.scss";
import { Link } from "react-router-dom";
import { IconComponent } from "@app/core/icon";
import classNames from "classnames";

export type ButtonVariant =
  | "primary-default"
  | "primary-brand"
  | "primary-inverted"
  | "secondary"
  | "secondary-inverted"
  | "tertiary"
  | "tertiary-inverted"
  | "link-primary"
  | "link-secondary"
  | "link-tertiary"
  | "link-quaternary";

export type IconStyle = "outline" | "filled" | "outline-fill";

export interface IClickableComponentProps {
  animated?: boolean;

  buttonType?: "submit" | "reset" | "button";
  disabled?: boolean;
  disableSelect?: boolean;
  dynamicSize?: boolean;
  fullWidth?: boolean;
  href?: string;
  iconFillColor?: string;
  iconLeft?: string;
  iconRight?: string;
  iconStyle?: IconStyle;
  onClick?: () => void;
  size?: number;
  target?: "_blank" | "_self";
  title?: string;
  variant?: ButtonVariant;
  zIndex?: number;
}

const renderIconMargin = (
  margin: string,
  title: string | undefined,
  icon: string | undefined,
  iconFillColor: string | undefined
) => {
  let style;
  if (title) {
    style = `${styles[`icon-${margin}`]} `;
  }

  return (
    <span className={style}>
      <IconComponent icon={icon} size="14px" fillColor={iconFillColor} />
    </span>
  );
};

const renderIconText = (
  iconLeft: string | undefined,
  iconRight: string | undefined,
  title: string | undefined,
  iconFillColor: string | undefined
) => {
  const showTextOnTheCenter = iconRight && iconLeft;

  return (
    <span className={styles["icon-svg"]}>
      {iconLeft && renderIconMargin(showTextOnTheCenter ? "center" : "left", title, iconLeft, iconFillColor)}
      {title}
      {iconRight && renderIconMargin(showTextOnTheCenter ? "center" : "right", title, iconRight, iconFillColor)}
    </span>
  );
};

const ClickableComponent = (props: IClickableComponentProps) => {
  const [iconText, setIconText] = React.useState<JSX.Element>(
    renderIconText(props.iconLeft, props.iconRight, props.title, props.iconFillColor)
  );
  useEffect(() => setIconText(renderIconText(props.iconLeft, props.iconRight, props.title, props.iconFillColor)), [
    props.iconLeft,
    props.iconRight,
    props.title
  ]);
  const { buttonType, disableSelect, onClick, href, variant, title, disabled, iconStyle, target, zIndex } = props;

  const classModify = variant || "primary-default";
  const buttonFAB = !title ? styles["button--FAB"] : "";
  const animated = props.animated ? styles["animated"] : "";
  const noSelect = disableSelect ? styles[`no-select`] : "";
  const iconOutline = iconStyle ? styles[`button--${classModify}--icon-${iconStyle}`] : "";
  const buttonClassName = classNames(
    styles["button"],
    styles[`button--${classModify}`],
    buttonFAB,
    animated,
    iconOutline,
    noSelect,
    {
      "button--fullWidth": props.fullWidth
    },
    {
      "button__dynamic-size": props.dynamicSize
    }
  );
  const buttonStyle = {
    width: props.fullWidth ? "100%" : props.size,
    height: props.size,
    zIndex
  };

  if (href) {
    const regex = "https://|http://";
    const isExternalLink = href.match(regex) !== null ? true : false;

    if (isExternalLink) {
      return (
        <a target={target || "_blank"} className={buttonClassName} style={buttonStyle} href={href}>
          {iconText}
        </a>
      );
    }

    return (
      <Link target={target || "_self"} className={buttonClassName} style={buttonStyle} to={href}>
        {iconText}
      </Link>
    );
  }

  return (
    <button
      disabled={disabled}
      style={buttonStyle}
      type={buttonType}
      name={title}
      className={buttonClassName}
      onClick={onClick}
    >
      <span className={styles["icon-svg"]}>{iconText}</span>
    </button>
  );
};

export { ClickableComponent };
