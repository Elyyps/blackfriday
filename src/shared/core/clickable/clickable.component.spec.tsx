import React from "react";
import { shallow } from "enzyme";
import { ClickableComponent } from "./clickable.component";
import IconDefault from "@assets/icons/chevron-right.svg";

describe("[Clickable]", () => {
  it("Should render component as button without crashing", () => {
    const renderedComponent = shallow(
      <ClickableComponent
        title="test"
        iconRight={IconDefault}
        iconStyle="outline-fill"
        variant={"primary-brand"}
        onClick={() => alert(`clicked button 2: 1`)}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
  it("Should render component as link without crashing (with animation and https link)", () => {
    const renderedComponent = shallow(
      <ClickableComponent
        title="test"
        iconRight={IconDefault}
        iconStyle="outline-fill"
        variant={"primary-brand"}
        href="https://www.youtube.com"
        fullWidth
        animated
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
  it("Should render component as link without crashing (with http link)", () => {
    const renderedComponent = shallow(
      <ClickableComponent
        title="test"
        iconRight={IconDefault}
        iconStyle="outline-fill"
        variant={"link-primary"}
        href="http://www.youtube.com"
        fullWidth
        animated
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
  it("Should render component as link without crashing (internal link)", () => {
    const renderedComponent = shallow(
      <ClickableComponent
        title="test"
        iconRight={IconDefault}
        iconStyle="outline-fill"
        variant={"link-primary"}
        href="/#hailtomfrontendking"
        fullWidth
        animated
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
