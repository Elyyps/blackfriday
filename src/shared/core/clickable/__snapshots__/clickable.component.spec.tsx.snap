// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Clickable] Should render component as button without crashing 1`] = `
<button
  className="button button--primary-brand button--primary-brand--icon-outline-fill"
  name="test"
  onClick={[Function]}
  style={
    Object {
      "height": undefined,
      "width": undefined,
      "zIndex": undefined,
    }
  }
>
  <span
    className="icon-svg"
  >
    <span
      className="icon-svg"
    >
      test
      <span
        className="icon-right "
      >
        <IconComponent
          icon="chevron-right.svg"
          size="14px"
        />
      </span>
    </span>
  </span>
</button>
`;

exports[`[Clickable] Should render component as link without crashing (internal link) 1`] = `
<Link
  className="button button--link-primary animated button--link-primary--icon-outline-fill button--fullWidth"
  style={
    Object {
      "height": undefined,
      "width": "100%",
      "zIndex": undefined,
    }
  }
  target="_self"
  to="/#hailtomfrontendking"
>
  <span
    className="icon-svg"
  >
    test
    <span
      className="icon-right "
    >
      <IconComponent
        icon="chevron-right.svg"
        size="14px"
      />
    </span>
  </span>
</Link>
`;

exports[`[Clickable] Should render component as link without crashing (with animation and https link) 1`] = `
<a
  className="button button--primary-brand animated button--primary-brand--icon-outline-fill button--fullWidth"
  href="https://www.youtube.com"
  style={
    Object {
      "height": undefined,
      "width": "100%",
      "zIndex": undefined,
    }
  }
  target="_blank"
>
  <span
    className="icon-svg"
  >
    test
    <span
      className="icon-right "
    >
      <IconComponent
        icon="chevron-right.svg"
        size="14px"
      />
    </span>
  </span>
</a>
`;

exports[`[Clickable] Should render component as link without crashing (with http link) 1`] = `
<a
  className="button button--link-primary animated button--link-primary--icon-outline-fill button--fullWidth"
  href="http://www.youtube.com"
  style={
    Object {
      "height": undefined,
      "width": "100%",
      "zIndex": undefined,
    }
  }
  target="_blank"
>
  <span
    className="icon-svg"
  >
    test
    <span
      className="icon-right "
    >
      <IconComponent
        icon="chevron-right.svg"
        size="14px"
      />
    </span>
  </span>
</a>
`;
